tfhe/src/core_crypto/experimental/algorithms/test/lwe_stair_keyswitch.rs

```diff
@@ -0,0 +1,296 @@
+            ks2_base_log,
+            ks2_level,
+            ks2_lwe_noise_distribution,
+            ciphertext_modulus,
+            &mut rsc.encryption_random_generator,
+        );
+
+        assert!(check_encrypted_content_respects_mod(
+            &ksk_inter_to_small,
+            ciphertext_modulus
+        ));
+
+        //PBS PART
+        let mut bsk = LweBootstrapKey::new(
+            Scalar::ZERO,
+            glwe_dimension.to_glwe_size(),
+            polynomial_size,
+            pbs_base_mpg,
+            pbs_level,
+            lwe_dimension,
+            ciphertext_modulus,
+        );
+
+        par_generate_lwe_bootstrap_key(
+            &small_lwe_secret_key,
+            &glwe_secret_key,
+            &mut bsk,
+            std_dev_bsk,
+            &mut rsc.encryption_random_generator,
+        );
+
+        assert!(check_encrypted_content_respects_mod(
+            &*bsk,
+            ciphertext_modulus
+        ));
+
+        let mut fbsk = FourierLweBootstrapKey::new(
+            small_lwe_secret_key.lwe_dimension(),
+            glwe_dimension.to_glwe_size(),
+            polynomial_size,
+            pbs_base_mpg,
+            pbs_level,
+        );
+
+        convert_standard_lwe_bootstrap_key_to_fourier(&bsk, &mut fbsk);
+        drop(bsk);
+
+        let accumulator = generate_accumulator(
+            polynomial_size,
+            glwe_dimension.to_glwe_size(),
+            msg_modulus.cast_into(),
+            ciphertext_modulus,
+            delta,
+            |x| x,
+        );
+
+        for _ in 0..NB_TESTS {
+            let plaintext = Plaintext(msg * delta);
+
+            //Encryption
+            let mut large_lwe = LweCiphertext::new(
+                Scalar::ZERO,
+                large_lwe_dimension.to_lwe_size(),
+                ciphertext_modulus,
+            );
+
+            encrypt_lwe_ciphertext(
+                &large_lwe_secret_key,
+                &mut large_lwe,
+                plaintext,
+                std_dev_bsk,
+                &mut rsc.encryption_random_generator,
+            );
+
+            assert!(check_encrypted_content_respects_mod(
+                &large_lwe,
+                ciphertext_modulus
+            ));
+
+            //Shrinking KS
+            let mut inter_lwe = LweCiphertext::new(
+                Scalar::ZERO,
+                inter_lwe_secret_key.lwe_dimension().to_lwe_size(),
+                ciphertext_modulus,
+            );
+
+            let mut small_lwe = LweCiphertext::new(
+                Scalar::ZERO,
+                lwe_dimension.to_lwe_size(),
+                ciphertext_modulus,
+            );
+
+            let mut out_pbs_ct = LweCiphertext::new(
+                Scalar::ZERO,
+                large_lwe_secret_key.lwe_dimension().to_lwe_size(),
+                ciphertext_modulus,
+            );
+
+            // Check the AP works for several iterations
+            for _ in 0..NB_TESTS {
```

Comment:
`for _ in 0..NB_TESTS` nested loops

Reply:
what about it ? it's a loop to run the atomic pattern in a looped fashion, so that the input is the previous output